source/simulator/assemblers/stokes.cc

```diff
@@ -78,17 +80,37 @@ namespace aspect
 
           const double JxW = scratch.finite_element_values.JxW(q);
 
-          for (unsigned int i = 0; i < stokes_dofs_per_cell; ++i)
-            for (unsigned int j = 0; j < stokes_dofs_per_cell; ++j)
-              if (scratch.dof_component_indices[i] ==
-                  scratch.dof_component_indices[j])
-                data.local_matrix(i, j) += ((2.0 * eta * (scratch.grads_phi_u[i]
-                                                          * scratch.grads_phi_u[j]))
-                                            + one_over_eta * pressure_scaling
-                                            * pressure_scaling
-                                            * (scratch.phi_p[i]
-                                               * scratch.phi_p[j]))
-                                           * JxW;
+          if (assemble_A_approximation)
+            {
+              for (unsigned int i = 0; i < stokes_dofs_per_cell; ++i)
+                for (unsigned int j = 0; j < stokes_dofs_per_cell; ++j)
+                  if (scratch.dof_component_indices[i] ==
+                      scratch.dof_component_indices[j])
+                    {
+                      data.local_matrix(i, j) += ((2.0 * eta * (scratch.grads_phi_u[i]
+                                                                * scratch.grads_phi_u[j]))
+                                                  + one_over_eta * pressure_scaling
+                                                  * pressure_scaling
+                                                  * (scratch.phi_p[i]
+                                                     * scratch.phi_p[j]))
+                                                 * JxW;
+                    }
+
+            }
+          else
+            {
+              for (unsigned int i = 0; i < stokes_dofs_per_cell; ++i)
+                for (unsigned int j = 0; j < stokes_dofs_per_cell; ++j)
+                  if (scratch.dof_component_indices[i] ==
+                      scratch.dof_component_indices[j])
+                    {
+                      data.local_matrix(i, j) += (one_over_eta * pressure_scaling
```

Comment:
this will still do a lot of unnecessary work because it loops over all velocity dofs and multiplies zeros. If you don't want to fix it (for example by looking at ``fe.system_to_component_index(i).first``), can you leave a TODO?